### ISSUE TITLE
Some self-excluding type constraint combinations are allowed, wrong error when they aren't

### ISSUE THREAD

**Issue by abelbraaksma** [dotnet/fsharp#9777]
If you use a constraint combination that isn't allowed, you always get the same error, twice:

> Program.fs(38,76): error FS0043: The constraints 'struct' and 'not struct' are inconsistent
> Program.fs(38,76): error FS0043: The constraints 'struct' and 'not struct' are inconsistent

![image](https://user-images.githubusercontent.com/16015770/88439475-a2d43600-ce0b-11ea-9e20-d06467640e94.png)

Repro:

```f#
type Del<'T when 'T: comparison and 'T: delegate<unit, unit>> =
    member _.Z = 1
```

Some others are allowed, but cannot be instantiated:

```f#
type EnumNull<'T when 'T:enum<int> and 'T:null>()
type StructNull<'T when 'T:'T:struct and 'T:null>()
type DelEnum<'T when 'T: enum<int> and 'T: delegate<unit, unit>>
type UnmDel<'T when 'T: unmanaged and 'T: delegate<unit, unit>> =
// etc
```

As was mentioned on Slack, some of these are used with SRTP dirty tricks (see [slack discussion](https://fsharp.slack.com/archives/C1R50TKEU/p1595595515248400) with @NinoFloris and an example: https://github.com/fsprojects/FSharpPlus/blob/7e32d52e9881ff40b5440f5ca886b56e6dd4a4c4/src/FSharpPlus/Control/Alternative.fs#L28)

### Proposal: fix

I propose we fix this:

* Fix the error to show the actual conflicting / inconsistent constraints
* Fix the error to show only once
* Make others that shouldn't be allowed (i.e., that cannot be combined in any type) a warning, and add it as default to `warnings-as-errors` (to prevent this from becoming a backward compat issue, similar to how we added FS3239 by default to new projects).

**Comment by gusty** [comment]
Since this is used as a feature in F#+ as you already pointed, this could be a breaking change, I mean it would stop compiling.

**Comment by TysonMN** [comment]
Quoting https://github.com/fsharp/fslang-suggestions/issues/349#issuecomment-705836005
> Most likely, if it gets fixed, it becomes a warning, and people can opt in making it an error.

Turning all of these cases into warnings seems like a great idea to me.

**Comment by gusty** [comment]
Yes, I think that's a good idea but we have to be careful as warnings affect overload resolution.

**Comment by TysonMN** [comment]
How do warnings affect overload resolution?

**Comment by abelbraaksma** [comment]
This issue was created for the generic type constraint, when such a constraint cannot be fulfilled, I don't think this has anything to do with overloads. Also, new warnings don't have any effect on existing code, unless we would also change behavior if a warning is raised. I don't think that'd be sensible to do.

@gusty, do you have an example where a warning changes overload resolution?

**Comment by TysonMN** [comment]
> Also, new warnings don't have any effect on existing code...

Quoting https://github.com/fsharp/fslang-suggestions/issues/349#issuecomment-705836005
> Most likely, if it gets fixed, it becomes a warning, and people can opt in making it an error.

Although I generally agree with you, I am also sympathetic to the argument that adding a warning is a breaking change for programs that already treat warnings as errors.  I think the [idea of `AnalysisLevel` that is being introduced with .NET 5](https://devblogs.microsoft.com/dotnet/automatically-find-latent-bugs-in-your-code-with-net-5) is a good compromise here.

**Comment by gusty** [comment]
I don't have an example right now but I remember having played with it.

If you look into the compiler's source code, you'll see the tie breaker looks for warnings.

Sorry for not having more information, what I say is, let's make sure it doesn't affect it.

**Comment by abelbraaksma** [comment]
That's interesting, I'll have a look at that in the code, I certainly wasn't aware and I'd be curious what specific warnings can lead to change of behavior (unless the warning itself is already a sign of the type checker detecting a fault, like defaulting to `obj` on generics).

@TysonMN, looking at previous discussions making an overlooked fault into either an error or a warning, the typical resolution is to go for a warning, sometimes adding that warning to new projects to the "warnings as errors" list. True, there's an off chance this leads to issues, but anyone treating all warnings as errors (rare, but possible) can use `#nowarn`, or configure this differently.

Since such warnings are introduced for cases that typically are wrong already anyway, in the rare case that code exploits this, they can switch the warning off in the offending file, or globally.

**Comment by TysonMN** [comment]
Completely agree.

> ...anyone treating all warnings as errors (rare, but possible)...

I am such a rarity :)  Actually, I only treat warnings as errors in release mode.  In debug mode, I allow developers the extra freedom to have warnings not prevent them from compiling, running tests, or running the program.

A "breaking change" is often discussed as something that must be avoided at all costs.  I don't like that perspective, because I don't think all breaking changes are equally bad.  The ones being considered here are good examples of ones that are not that bad.
